uqcsbot/api.py

```diff
@@ -133,4 +138,137 @@ def __call__(self, **kwargs) -> 'APIWrapper':
         """
         On call, reconfigure the API Wrapper
         """
-        return type(self)(self._client, **kwargs)
\ No newline at end of file
+        return type(self)(self._client, **kwargs)
+
+
+class Channel(object):
+    def __init__(
+        self,
+        bot: 'UQCSBot',
+        channel_id: str,
+        name: str,
+        is_public: bool = False,
+        is_private: bool = False,
+        is_archived: bool = False,
+        previous_names: List[str] = None
+    ): 
+        self._bot = bot
+        self.id = channel_id
+        self.name = name
+        self._members = None
+        self.is_public = is_public
+        self.is_private = is_private
+        self.is_archived = is_archived
+        self.previous_names = previous_names or []
+
+    @property
+    def members(self) -> List[str]:
+        if self._members is not None:
+            return self._members
+        members = []
+        for page in self._bot.api.conversations.members.paginate(channel=self.id):
+            members += page["members"]
+        self._members = members
+        return self._members
+
+    def update_members(self) -> List[str]:
+        self._members = None
+        return self.members
+
+
+class ChannelWrapper(object):
+    def __init__(self, bot: 'UQCSBot'):
+        self._bot = bot
+        self._initialised = False
+        self._channels_by_id = {}
+        self._channels_by_name = {}
+        self._lock = threading.RLock()
+        self._bind_handlers()
+
+    def _bind_handlers(self) -> None:
+        PREFIX = "_on_"
+        for name in dir(type(self)):
+            if not name.startswith(PREFIX):
+                continue
+            attr = getattr(self, name)
+            mtype = name[len(PREFIX):]
+            self._bot.on(mtype, attr)
+
+    def _add_channel(self, chan_dict: dict) -> Channel:
+        chan = Channel(
+            bot=_bot,
+            channel_id=chan_dict['id'],
+            name=chan_dict['name'],
+            is_public=chan_dict.get('is_public', False),
+            is_private=chan_dict.get('is_private', False),
+            is_archived=chan_dict.get('is_archived', False),
+        )
+        self._channels_by_name[chan.name] = chan
+        self._channels_by_id[chan.id] = chan
+        return chan
+
+    def _initialise(self):
+        with self._lock:
+            for page in self._bot.api.channels.list.paginate():
+                for chan in page['channels']:
+                    self._add_channel(chan)
+            self._initialised = True
+
+    def get(self, name_or_id: str, default: T = None, use_cache: bool =True) -> Union[Channel, T]:
```

Comment:
I'd rather have two named parameters, name and id than a joint variable here I think. This also means that we can throw do something different depending on if they passed in name or id when they said use_cache = False

Reply:
I'm trying to replicate `dict` behaviour here with `.get`

Reply:
Hmm, fair enough. Maybe add another method then called `invalidate_caches` so that the user has someway of forcing it to reload data for channels by name.